### PR TITLE
[ENG-1369][ENG-1368][ENG-1367] Sloan Icons Omnibus

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Since the Sloan icons PRs have been mostly aproved individually but will conflict if merged one-by-one I've created this PR to combine and standardize the code into one consistent PR.

## Changes

- adds new data/prereg/coi based fields to the preprint model
- adds new data/prereg/coi update methods to the preprint model
- adds new data/prereg/coi fields to serialize with customization that validates for switch activeness.
- add new tests for all icons.

## QA Notes

There is a small migration, but no user facing changes at this point. Dev test only.

## Documentation

A lot of docstrings, but noting user facing

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-1367
https://openscience.atlassian.net/browse/ENG-1368
https://openscience.atlassian.net/browse/ENG-1369

### CODE REVIEW

File: osf/models/preprint.py [hunk 1]
```diff
@@ -925,6 +979,298 @@ def create_waterbutler_log(self, auth, action, payload):
             params=params
         )
 
+    def update_has_coi(self, auth: tuple, has_coi: bool, log: bool = True, save: bool = True):
```
Comment (brianjgeiger) [inline osf/models/preprint.py]: Look who's gotten all fancy

Comment (Johnetordoff) [reply osf/models/preprint.py]: Python 3 is classy language

File: osf/models/preprint.py [hunk 2]
```diff
@@ -925,6 +979,298 @@ def create_waterbutler_log(self, auth, action, payload):
...
+        :param save: Boolean should this be saved immediately?
+        :return:
+
+        This method brought to you via a grant from the Alfred P Sloan Foundation.
+        """
+        if self.conflict_of_interest_statement == coi_statement:
+            return
+
+        if not self.has_coi:
+            raise ValidationError('You do not have ability to edit a conflict of interest while the has_coi field is '
```
Comment (aaxelb) [inline osf/models/preprint.py]: `ValidationError` seems like the wrong exception -- this has nothing to do with django's validation

our own `PreprintStateError` seems more appropriate

File: osf/models/preprint.py [hunk 3]
```diff
@@ -925,6 +979,298 @@ def create_waterbutler_log(self, auth, action, payload):
...
+        :param save: Boolean should this be saved immediately?
+        :return:
+
+        This method brought to you via a grant from the Alfred P Sloan Foundation.
+        """
+        if self.data_links == data_links:
+            return
+
+        if not self.has_data_links:
+            raise ValidationError('You cannot edit this statement while your data links availability is set to false or'
```
Comment (aaxelb) [inline osf/models/preprint.py]: `PreprintStateError`

File: osf/models/preprint.py [hunk 4]
```diff
@@ -925,6 +979,298 @@ def create_waterbutler_log(self, auth, action, payload):
...
+
+        This method brought to you via a grant from the Alfred P Sloan Foundation.
+        """
+        if self.why_no_data == why_no_data:
+            return
+
+        if self.has_data_links is False:
+            self.why_no_data = why_no_data
+        else:
+            raise ValidationError('You cannot edit this statement while your data links availability is set to true or'
```
Comment (aaxelb) [inline osf/models/preprint.py]: `PreprintStateError`

File: osf/models/preprint.py [hunk 5]
```diff
@@ -925,6 +979,298 @@ def create_waterbutler_log(self, auth, action, payload):
...
+        :param save: Boolean should this be saved immediately?
+        :return:
+
+        This method brought to you via a grant from the Alfred P Sloan Foundation.
+        """
+        if why_no_prereg == self.why_no_prereg:
+            return
+
+        if self.has_prereg_links or self.has_prereg_links is None:
+            raise ValidationError('You cannot edit this statement while your prereg links '
```
Comment (aaxelb) [inline osf/models/preprint.py]: `PreprintStateError`

File: osf/models/preprint.py [hunk 6]
```diff
@@ -925,6 +979,298 @@ def create_waterbutler_log(self, auth, action, payload):
...
+        :param save: Boolean should this be saved immediately?
+        :return:
+
+        This method brought to you via a grant from the Alfred P Sloan Foundation.
+        """
+        if self.prereg_link_info == prereg_link_info:
+            return
+
+        if not self.has_prereg_links:
+            raise ValidationError('You cannot edit this field while your prereg links'
```
Comment (aaxelb) [inline osf/models/preprint.py]: `PreprintStateError`

File: api_tests/preprints/views/test_preprint_detail.py [hunk 7]
```diff
@@ -752,6 +759,367 @@ def test_update_preprint_task_called_on_api_update(
 
         assert mock_on_preprint_updated.called
 
+    def test_update_has_coi(self, app, user, preprint, url):
```
Comment (aaxelb) [inline api_tests/preprints/views/test_preprint_detail.py]: nice tests 🥇

**Comment by Johnetordoff** [comment]
merged into https://github.com/CenterForOpenScience/osf.io/pull/9291

**Pull request closed without merging**